src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -55,11 +51,20 @@ def __init__(
         self.preparer = preparer
         self._python_candidate = RequiresPythonCandidate(py_version_info)
         self._make_install_req_from_spec = make_install_req
-        self._ignore_installed = ignore_installed
+        self._force_reinstall = force_reinstall
         self._ignore_requires_python = ignore_requires_python
+
         self._link_candidate_cache = {}  # type: Cache[LinkCandidate]
         self._editable_candidate_cache = {}  # type: Cache[EditableCandidate]
 
+        if not ignore_installed:
+            self._installed_dists = {
+                dist.project_name: dist
+                for dist in get_installed_distributions()
+            }
+        else:
+            self._installed_dists = {}
```

Comment:
We can instantiate `self._installed_dists = {}` above with other dictionary instantions at line 62. 
This will also get rid of the `else` condition

Reply:
This would require a separate `# type:` comment since Mypy is not smart enough to infer types with scoping rules, and also trigger an unnecessary assignment at runtime. I prefer this to do it like this personally.